pkg/util/strategicpatch/patch.go

```diff
@@ -31,7 +31,352 @@ import (
 // lists should be merged or replaced.
 //
 // For more information, see the PATCH section of docs/api-conventions.md.
+
+// CreateStrategicMergePatch creates a patch that can be passed to StrategicMergePatch.
```

Comment:
Let's walk through an example.

```
Original object stored as annotation:
containers:
  -  name: a
     image: x
  -  name: b
     image: y

New local object to reconcile:
containers:
  -  name: b
     image: y
     settings:
       flag1: true

Current server object:
containers:
  -  name: a
     image: x
  -  name: b
     image: y
     settings:
       flag2: true
  -  name: c
     image: z

Patch we should generate:
containers:
  -  name: a
     $patch: delete
  -  name: b
     image: y
     settings:
       flag1: true
```

I am going to assume that, not only is the above correct, but that the above situation would not cause a conflict. flag1 and flag2 would simply merge together and both be set to "true". There are two situations I can think of where the data would be such that we could throw an error, if we want it to:

1. If "Current server object" had some information added to container "a", then we could decide whether that throws a conflict. I think it would be safe to say that it doesn't. It also would be hard to throw an error because it can't be deduced just by looking at the patch and the server object.
2. If "Current server object" didn't have flag1 set to true, but instead flag2 set to false. That would represent a true irreconcilable conflict. So we could either force-clobber, or we could try to detect and warn. If we do want to detect, as @jackgr says, I think we could do so without needing a non-additions API and instead detect by generating the patch and comparing it to the live server object.

In other words I think I'm agreeing with @jackgr that we don't need the additions API and a comparison of the live server object with the patch should be sufficient. @bgrant0607 can you outline a scenario where that is not sufficient?

@jackgr re: incompatible change constraints, should that kind of thing not just be implemented in the server-side validation of the final state of the objects post-patch, instead of being re-implemented and re-checked in this library?